### ISSUE TITLE
Change TFS URL 

### ISSUE THREAD

**Issue by bredar** [tfsaggregator/tfsaggregator#115]
Hello

Somehow, connections from the tfs server locally, via https, can't authenticate. Log on to tfs via http://localhost:8080/tfs works, but via https://tfs.me.xyz/tfs won't. Seems to be an IIS problem.

tfsaggregator uses the public address to connect. So, if I could change the tfs url to http://localhost:8080/tfs, the problem would be solved. I think, it was possible to change the url in version 1. Is there any way to change it in the actual version?

Regards

**Comment by giuliov** [comment]
No, this is not possible today.
This version asks TFS using the `ILocationDataProvider.GetSelfReferenceUri` method, which solve other issues, for example using the Collection matching the request.

Can you report the URLs displayed in the Console?
![2016-03-01_192749](https://cloud.githubusercontent.com/assets/1869724/13437346/0e3ec7d6-dfe4-11e5-917c-491c9160fab2.png)

**Comment by bredar** [comment]
Hello

Thanks for your answer. I’ve configured this settings (see attachment)

Regards

Francois




Von: Giulio Vian [mailto:notifications@github.com]
Gesendet: Dienstag, 1. März 2016 19:31
An: tfsaggregator/tfsaggregator <tfsaggregator@noreply.github.com>
Cc: François Zbinden <zbinden@dataflow-solutions.ch>
Betreff: Re: [tfsaggregator] Change TFS URL (#115)


No, this is not possible today.
This version asks TFS using the ILocationDataProvider.GetSelfReferenceUri method, which solve other issues, for example using the Collection matching the request.

Can you report the URLs displayed in the Console?
[2016-03-01_192749]<https://cloud.githubusercontent.com/assets/1869724/13437346/0e3ec7d6-dfe4-11e5-917c-491c9160fab2.png>

**Comment by giuliov** [comment]
GitHub drops email attachments: you should use the web interface to insert pictures.

**Comment by bredar** [comment]
Here you go.
![tfs settings](https://cloud.githubusercontent.com/assets/17574488/13500013/efc8ac1c-e160-11e5-9906-9196da796c7f.PNG)

**Comment by bredar** [comment]
Btw I see in the event log, he is not able to create the connection:
`Detailed Message: Exception encountered processing notification: TF30063: You are not authorized to access https://tfs.*****.ch/tfs/DFCollection. `

**Comment by giuliov** [comment]
Some things to check.

- The SSL configuration is using the same application pool as port 8080?
- The service account running TFS (better, the account running the SSL application pool) has access to the Collection?

You can also look at the [Troubleshooting](https://github.com/tfsaggregator/tfsaggregator/wiki/Troubleshooting) page in the documentation.

**Comment by fishbits** [comment]
I'm encountering a similar issue, but I'm not running SSL...
```
[1228] TFSAggregator: [Critical]    Exception encountered processing notification: TF30063: You are not authorized to access http://[friendlyDNS-tfs-url]:8080/tfs/[collectionName]. 
[1228]    at Aggregator.Core.Facade.WorkItemRepository..ctor(Uri tfsCollectionUri, IdentityDescriptor toImpersonate, ILogEvents logger) in E:\Git\tfsaggregator\Aggregator.Core\Facade\WorkItemRepository.cs:line 38 
[1228]    at TFSAggregator.TfsSpecific.WorkItemChangedEventHandler.<>c.<ProcessEvent>b__2_0(Uri collectionUri, IdentityDescriptor toImpersonate, ILogEvents logEvents) in E:\Git\tfsaggregator\Aggregator.ServerPlugin\WorkItemChangedEventHandler.cs:line 68 
[1228]    at Aggregator.Core.Context.RuntimeContext.GetWorkItemRepository() in E:\Git\tfsaggregator\Aggregator.Core\Context\RuntimeContext.cs:line 157 
[1228]    at Aggregator.Core.EventProcessor..ctor(IRuntimeContext runtime) in E:\Git\tfsaggregator\Aggregator.Core\EventProcessor.cs:line 40 
[1228]    at TFSAggregator.TfsSpecific.WorkItemChangedEventHandler.ProcessEvent(IVssRequestContext requestContext, NotificationType notificationType, Object notificationEventArgs, Int32& statusCode, String& statusMessage, ExceptionPropertyCollection& properties) in E:\Git\tfsaggregator\Aggregator.ServerPlugin\WorkItemChangedEventHandler.cs:line 84
```
This is based on a local build I did of master, Debug-2015.1 (hence the local drive references)

My configuration is:
![image](https://cloud.githubusercontent.com/assets/5439831/13831247/7cd47f68-eb8f-11e5-90d6-626d1f6a6dd7.png)

We have two application tier servers in a Big-IP-based load balanced pool, associated with a VIP that's attached to a friendly DNS name.

The desired change occurs when the policy conditions are met -- I've tested and verified with the console app and directly on the task board.  I just have to manually refresh the affected work item (a PBI in this case) to see the change.

In my configuration above, the friendly DNS name is used in all three URLs. I'm wondering if each application tier's 'Server URL' shouldn't be using 'localhost'? I inherited the configuration this way, and I cannot seem to find any load balanced examples that *dont* use 'localhost' for the Server URL.

Btw, the exception above is typically preceded by a list of successful rule application events on one application tier, with the exception occurring on the other application tier.

Thanks!

**Comment by giuliov** [comment]
I do not disdain the idea of enabling an administrator to force a URL, but it is not that easy to squeeze in existing code: it tries avoiding parsing URL because:

- TFS can be on a different port / virtual dir
- there can be more bindings with different hostnames
- HTTP vs HTTPS
- VSTS vs TFS URL
- VSTS evolutions (e.g. <https://blogs.msdn.microsoft.com/visualstudioalm/2016/01/11/how-we-plan-to-enable-creating-multiple-collections-per-account/>)

Hope to produce a solution soon.

@fishbits You are correct that Server URL should simply refer to `localhost`, see https://blogs.msdn.microsoft.com/tfsao/2010/04/07/server-and-notification-url/

**Comment by jessehouwing** [comment]
Are you guys certain that the TFS Service account is at least contributor to the team project collection it's trying to update?

**Comment by giuliov** [comment]
@fishbits may you try PR #120 if it works for you?

**Comment by bredar** [comment]
Just to inform you, in my configration, localhost and my public tfs url are provided by the same virtual directory, and therefore using the exact same configuration. But thanks for you pull.

**Comment by fishbits** [comment]
@jessehouwing The service account being used in our instance has many Administrative privileges (including project collection admin), but it is not, specifically, a Contributor in the team project in question.

@giuliov Just tested in my sandbox, and then inserted into production.  Now, instead of trying to hit our load-balanced entry point, we get the following:

`[2000] TFSAggregator: [Diagnostic]  Built a new Work Item Repository for http://localhost:8080/tfs/[collectionName]/Services/v3.0/LocationService.asmx`

Looks successful so far; no exceptions to report at this time. Should see more activity on our boards in the morning (US PDT). 

Note that prior to this, I did not have the opportunity to test our production environment with a changed ServerURL setting to 'localhost'. So, the only change made & tested at this time was this PR. Thanks, @giuliov!

**Comment by fishbits** [comment]
Did run into some exceptions overnight, but only on one task. It did not have a parent and was related to several changesets. I'm not sure why it did not just bypass my policies and continue...

```
[2000] TFSAggregator: [Diagnostic]  Built a new Work Item Repository for http://localhost:8080/tfs/[ProjectCollectionName]/Services/v3.0/LocationService.asmx 
[2000] TFSAggregator: [Verbose]     Building Script Engine for C# 
[2000] TfsAggregator.ServerPlugin Information: 12767 : 
[2000] TFSAggregator: [Information] Starting processing on workitem #118633 
[2000] TFSAggregator: [Diagnostic]  Policy scope Collections([ProjectCollectionName]) matches [ProjectCollectionName] 
[2000] TFSAggregator: [Verbose]     Policy 'DefaultPolicy' applies 
[2000] TFSAggregator: [Verbose]     Evaluating Rule 'AutoCloseParent' 
[2000] TFSAggregator: [Diagnostic]  Rule scope WorkItemTypes(Task) matches Task 
[2000] TFSAggregator: [Verbose]     Applying Rule 'AutoCloseParent' on #118633 
[2000] TfsAggregator.ServerPlugin Critical: 47513 : 
[2000]    at Aggregator.Core.Script.DotNetScriptEngine`1.RunScript(Assembly assembly, String scriptName, IWorkItem self, IWorkItemRepository store) in E:\Git\tfsaggregator\Aggregator.Core\Script\DotNetScriptEngine{TCodeDomProvider}.cs:line 159 
[2000]    at Aggregator.Core.Script.DotNetScriptEngine`1.Run(String scriptName, IWorkItem workItem, IWorkItemRepository store) in E:\Git\tfsaggregator\Aggregator.Core\Script\DotNetScriptEngine{TCodeDomProvider}.cs:line 212 
[2000]    at Aggregator.Core.EventProcessor.ApplyRule(Rule rule, IWorkItem workItem) in E:\Git\tfsaggregator\Aggregator.Core\EventProcessor.cs:line 109 
[2000]    at Aggregator.Core.EventProcessor.ApplyRules(IWorkItem workItem, IEnumerable`1 rules) in E:\Git\tfsaggregator\Aggregator.Core\EventProcessor.cs:line 95 
[2000]    at Aggregator.Core.EventProcessor.ProcessEvent(IRequestContext requestContext, INotification notification) in E:\Git\tfsaggregator\Aggregator.Core\EventProcessor.cs:line 63 
[2000]    at TFSAggregator.TfsSpecific.WorkItemChangedEventHandler.ProcessEvent(IVssRequestContext requestContext, NotificationType notificationType, Object notificationEventArgs, Int32& statusCode, String& statusMessage, ExceptionPropertyCollection& properties) in E:\Git\tfsaggregator\Aggregator.ServerPlugin\WorkItemChangedEventHandler.cs:line 87 
[2000] TFSAggregator: [Critical]    Exception encountered processing notification: Object reference not set to an instance of an object. 
[2000]    at Aggregator.Core.Script.DotNetScriptEngine`1.RunScript(Assembly assembly, String scriptName, IWorkItem self, IWorkItemRepository store) in E:\Git\tfsaggregator\Aggregator.Core\Script\DotNetScriptEngine{TCodeDomProvider}.cs:line 159 
[2000]    at Aggregator.Core.Script.DotNetScriptEngine`1.Run(String scriptName, IWorkItem workItem, IWorkItemRepository store) in E:\Git\tfsaggregator\Aggregator.Core\Script\DotNetScriptEngine{TCodeDomProvider}.cs:line 212 
[2000]    at Aggregator.Core.EventProcessor.ApplyRule(Rule rule, IWorkItem workItem) in E:\Git\tfsaggregator\Aggregator.Core\EventProcessor.cs:line 109 
[2000]    at Aggregator.Core.EventProcessor.ApplyRules(IWorkItem workItem, IEnumerable`1 rules) in E:\Git\tfsaggregator\Aggregator.Core\EventProcessor.cs:line 95 
[2000]    at Aggregator.Core.EventProcessor.ProcessEvent(IRequestContext requestContext, INotification notification) in E:\Git\tfsaggregator\Aggregator.Core\EventProcessor.cs:line 63 
[2000]    at TFSAggregator.TfsSpecific.WorkItemChangedEventHandler.ProcessEvent(IVssRequestContext requestContext, NotificationType notificationType, Object notificationEventArgs, Int32& statusCode, String& statusMessage, ExceptionPropertyCollection& properties) in E:\Git\tfsaggregator\Aggregator.ServerPlugin\WorkItemChangedEventHandler.cs:line 87 
```

**Comment by jessehouwing** [comment]
What was the exact rule? If the rile to any to navigate to the parent it
should check for the parents presence before navigating. It can do so
using  self.HasParent .

**Comment by fishbits** [comment]
Attached is my policy file...
[TFSAggregator2.ServerPlugin.policies.txt](https://github.com/tfsaggregator/tfsaggregator/files/184759/TFSAggregator2.ServerPlugin.policies.txt)

**Comment by jessehouwing** [comment]
The first line calling self.parent.typename will throw a
NullReferenceException... So you'll want to start with a quick check
whether a parent is present and if not quickly return without further
processing.

**Comment by fishbits** [comment]
Thanks, @jessehouwing. Just wrapped each policy with a brief evaluation of:
`if (self.HasParent())`
...and the exception disappears.

**Comment by jessehouwing** [comment]
I had the same "permission" error after installing the latest build on TFS 2015 update 2, this is what I did to fix it:

```
C:\WINDOWS\system32>tfssecurity /collection:http://jessehouwing:8080/tfs/DefaultCollection /g+ "Project Collection administrators" "LOCAL SERVICE"
Microsoft (R) TFSSecurity - Team Foundation Server Security Tool
Copyright (c) Microsoft Corporation.  All rights reserved.

The target Team Foundation Server is http://jessehouwing:8080/tfs/defaultcollection.
Resolving identity "Project Collection administrators"...
a [A] [DefaultCollection]\Project Collection Administrators
Resolving identity "LOCAL SERVICE"...
  [U] NT AUTHORITY\LOCAL SERVICE
Adding LOCAL SERVICE to [DefaultCollection]\Project Collection Administrators...
Verifying...

SID: S-1-9-1551374245-1204400969-2402986413-2179408616-0-0-0-0-1

DN:

Identity type: Team Foundation Server application group
   Group type: AdministrativeApplicationGroup
Project scope: Server scope
 Display name: [DefaultCollection]\Project Collection Administrators
  Description: Members of this application group can perform all privileged operations on the Team Project Collection.

4 member(s):
a [A] [TEAM FOUNDATION]\Team Foundation Administrators
  [U] JESSEHOUWING\jhouw (Jesse Houwing)
  [U] NT AUTHORITY\LOCAL SERVICE
s [A] [DefaultCollection]\Project Collection Service Accounts

Member of 2 group(s):
e [A] [DefaultCollection]\Project Collection Valid Users
a [A] [Agent Queues]\Agent Queue Administrators

Done.
```

**Comment by jessehouwing** [comment]
Merged to Release/2.1.0-rc1
